pkg/ims/service/service.go

```diff
@@ -0,0 +1,111 @@
+package service
+
+import (
+	"crypto/rand"
+	"math/big"
+	"strconv"
+
+	"github.com/omc-college/management-system/pkg/ims/models"
+	"github.com/omc-college/management-system/pkg/ims/repository/postgresql"
+	"github.com/omc-college/management-system/pkg/pwd"
+)
+
+// token returns random int for verification token
+func token() string {
+	r, _ := rand.Int(rand.Reader, big.NewInt(int64(999999999999999)))
+	rand := strconv.Itoa(int(r.Int64()))
+	return rand
+}
+
+type SignUpService struct {
+	SignUpRepo 	*postgresql.SignUpRepository
+	CredRepo	*postgresql.CredRepository
+	UserRepo	*postgresql.UsersRepository
+}
+
+func NewSignUpService(repo *postgresql.SignUpRepository,credrepo *postgresql.CredRepository, userrepo *postgresql.UsersRepository) *SignUpService {
+	return &SignUpService{
+		SignUpRepo: repo,
+		CredRepo: credrepo,
+		UserRepo: userrepo,
+	}
+}
+
+func (service *SignUpService) SignUp (request *models.SignupRequest) error {
+	var cred models.Credentials
+	var tok models.EmailVerificationTokens
+	var err error
+	var user models.User
+
+	user.FirstName = request.FirstName
+	user.LastName = request.LastName
+	user.Email = request.Email
+
+	tok.VerificationToken = token()
+
+	cred.Salt = pwd.Salt(256 - len(request.Password))
+
+	cred.PasswordHash, err = pwd.Hash(request.Password, cred.Salt)
+	if err != nil {
+		return err
+	}
+
+	err = service.SignUpRepo.InsertUser(&user)
```

Comment:
you need to start a transaction before making your first request to DB

Reply:
I do not understand enough what do you mean

Reply:
The problem was resolved at the zoom conference.